packages/python/plotly/plotly/basedatatypes.py

```diff
@@ -3411,6 +3678,206 @@ def _index_is(iterable, val):
+                # specified, so we specify them here so the following routines can work
+                # (they need to append " domain" to xref or yref)
+                self.layout[layout_obj][-1].update(xref="x", yref="y")
+            new_layout_objs = tuple(
+                filter(
+                    lambda x: x is not None,
+                    [
+                        self._make_axis_spanning_layout_object(
+                            direction, self.layout[layout_obj][n],
+                        )
+                        for n in range(n_layout_objs_before, n_layout_objs_after)
+                    ],
+                )
+            )
+            self.layout[layout_obj] = (
+                self.layout[layout_obj][:n_layout_objs_before] + new_layout_objs
+            )
+
+    def add_vline(
+        self,
+        x,
+        row="all",
+        col="all",
+        exclude_empty_subplots=True,
+        annotation=None,
+        **kwargs
+    ):
+        self._process_multiple_axis_spanning_shapes(
+            dict(type="line", x0=x, x1=x, y0=0, y1=1),
+            row,
+            col,
+            "vline",
+            exclude_empty_subplots=exclude_empty_subplots,
+            annotation=annotation,
+            **kwargs
+        )
+        return self
+
+    add_vline.__doc__ = _axis_spanning_shapes_docstr("vline")
+
+    def add_hline(self, y, row="all", col="all", exclude_empty_subplots=True, **kwargs):
+        self._process_multiple_axis_spanning_shapes(
+            dict(type="line", x0=0, x1=1, y0=y, y1=y,),
+            row,
+            col,
+            "hline",
+            exclude_empty_subplots=exclude_empty_subplots,
+            **kwargs
+        )
+        return self
+
+    add_hline.__doc__ = _axis_spanning_shapes_docstr("hline")
+
+    def add_vrect(
+        self, x0, x1, row="all", col="all", exclude_empty_subplots=True, **kwargs
+    ):
+        self._process_multiple_axis_spanning_shapes(
+            dict(type="rect", x0=x0, x1=x1, y0=0, y1=1),
+            row,
+            col,
+            "vrect",
+            exclude_empty_subplots=exclude_empty_subplots,
+            **kwargs
+        )
+        return self
+
+    add_vrect.__doc__ = _axis_spanning_shapes_docstr("vrect")
+
+    def add_hrect(
+        self, y0, y1, row="all", col="all", exclude_empty_subplots=True, **kwargs
+    ):
+        self._process_multiple_axis_spanning_shapes(
+            dict(type="rect", x0=0, x1=1, y0=y0, y1=y1),
+            row,
+            col,
+            "hrect",
+            exclude_empty_subplots=exclude_empty_subplots,
+            **kwargs
+        )
+        return self
+
+    add_hrect.__doc__ = _axis_spanning_shapes_docstr("hrect")
+
+    def _has_subplots(self):
+        """ Returns True if figure contains subplots, otherwise it contains a
+        single plot and so this returns False. """
+        return self._grid_ref is not None
+
+    def _subplot_contains_trace(self, xref, yref):
```

Comment:
Right now a subplot with no traces but with a shape *does* render and "isn't empty" so I'm sort of thinking this should extend to `shapes` and `annotations` and `images` as well...

Reply:
you mean we should use a function `_subplot_contains_shape` or something like that? Only add to plots with shapes already on them?